Import Pandas Library to handle Data.

In [1]:
import pandas as pd

Save the file path in a variable.

In [2]:
file_path = 'melb_data.csv'

Use pd.read_csv to extract the data and store it inside a Variable as a DataFrame.

In [22]:
house_data = pd.read_csv(file_path)

Check the first five rows of the DataFrame.

In [23]:
house_data.head()

Suburb           Address  Rooms Type      Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

Notice that as per our directions, Python has only shown first five rows of the Data Set. Also notice that there are a total of 21 Columns.

The .describe() method in Pandas selects the Columns with Numerical Datatypes and shows the mean, spread, min & max values and count.

In [5]:
house_data.describe()

Rooms         Price      Distance      Postcode      Bedroom2  \
count  13580.000000  1.358000e+04  13580.000000  13580.000000  13580.000000   
mean       2.937997  1.075684e+06     10.137776   3105.301915      2.914728   
std        0.955748  6.393107e+05      5.868725     90.676964      0.965921   
min        1.000000  8.500000e+04      0.000000   3000.000000      0.000000   
25%        2.000000  6.500000e+05      6.100000   3044.000000      2.000000   
50%        3.000000  9.030000e+05      9.200000   3084.000000      3.000000   
75%        3.000000  1.330000e+06     13.000000   3148.000000      3.000000   
max       10.000000  9.000000e+06     48.100000   3977.000000     20.000000   

           Bathroom           Car       Landsize  BuildingArea    YearBuilt  \
count  13580.000000  13518.000000   13580.000000   7130.000000  8205.000000   
mean       1.534242      1.610075     558.416127    151.967650  1964.684217   
std        0.691712      0.962634    3990.669241    541.014538    37.273762   
min        0.000000      0.000000       0.000000      0.000000  1196.000000   
25%        1.000000      1.000000     177.000000     93.000000  1940.000000   
50%        1.000000      2.000000     440.000000    126.000000  1970.000000   
75%        2.000000      2.000000     651.000000    174.000000  1999.000000   
max        8.000000     10.000000  433014.000000  44515.000000  2018.000000   

          Lattitude    Longtitude  Propertycount  
count  13580.000000  13580.000000   13580.000000  
mean     -37.809203    144.995216    7454.417378  
std        0.079260      0.103916    4378.581772  
min      -38.182550    144.431810     249.000000  
25%      -37.856822    144.929600    4380.000000  
50%      -37.802355    145.000100    6555.000000  
75%      -37.756400    145.058305   10331.000000  
max      -37.408530    145.526350   21650.000000

Notice that count in Rooms and count in Year Built Column (and also some other Column) is different. This is because some values are missing i.e. Data is not entered. This causes the entire entry (row) to be effected. We may find the total rows before deleting the rows having missing entries!

In [25]:
house_data.shape

(13580, 21)

There are a total of 13580 Rows and the usual 21 Columns in the Data set. We may remove the remove the rows with missing values and find the total rows again!

In [26]:
house_data = house_data.dropna(axis=0)

In [27]:
house_data.shape

(6196, 21)

Notice that number of rows in the data set has been reduced from 13580 to 6196. More than half of the rows have been deleted!

In [28]:
house_data.head()

Suburb          Address  Rooms Type      Price Method SellerG  \
1  Abbotsford  25 Bloomburg St      2    h  1035000.0      S  Biggin   
2  Abbotsford     5 Charles St      3    h  1465000.0     SP  Biggin   
4  Abbotsford      55a Park St      4    h  1600000.0     VB  Nelson   
6  Abbotsford     124 Yarra St      3    h  1876000.0      S  Nelson   
7  Abbotsford    98 Charles St      2    h  1636000.0      S  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   
6  7/05/2016       2.5    3067.0  ...       2.0  0.0     245.0         210.0   
7  8/10/2016       2.5    3067.0  ...       1.0  2.0     256.0         107.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   
6     1910.0        Yarra  -37.8024    144.9993  Northern Metropolitan   
7     1890.0        Yarra  -37.8060    144.9954  Northern Metropolitan   

  Propertycount  
1        4019.0  
2        4019.0  
4        4019.0  
6        4019.0  
7        4019.0  

[5 rows x 21 columns]

In [29]:
house_data.describe()

Rooms         Price     Distance     Postcode     Bedroom2  \
count  6196.000000  6.196000e+03  6196.000000  6196.000000  6196.000000   
mean      2.931407  1.068828e+06     9.751097  3101.947708     2.902034   
std       0.971079  6.751564e+05     5.612065    86.421604     0.970055   
min       1.000000  1.310000e+05     0.000000  3000.000000     0.000000   
25%       2.000000  6.200000e+05     5.900000  3044.000000     2.000000   
50%       3.000000  8.800000e+05     9.000000  3081.000000     3.000000   
75%       4.000000  1.325000e+06    12.400000  3147.000000     3.000000   
max       8.000000  9.000000e+06    47.400000  3977.000000     9.000000   

          Bathroom          Car      Landsize  BuildingArea    YearBuilt  \
count  6196.000000  6196.000000   6196.000000   6196.000000  6196.000000   
mean      1.576340     1.573596    471.006940    141.568645  1964.081988   
std       0.711362     0.929947    897.449881     90.834824    38.105673   
min       1.000000     0.000000      0.000000      0.000000  1196.000000   
25%       1.000000     1.000000    152.000000     91.000000  1940.000000   
50%       1.000000     1.000000    373.000000    124.000000  1970.000000   
75%       2.000000     2.000000    628.000000    170.000000  2000.000000   
max       8.000000    10.000000  37000.000000   3112.000000  2018.000000   

         Lattitude   Longtitude  Propertycount  
count  6196.000000  6196.000000    6196.000000  
mean    -37.807904   144.990201    7435.489509  
std       0.075850     0.099165    4337.698917  
min     -38.164920   144.542370     389.000000  
25%     -37.855438   144.926198    4383.750000  
50%     -37.802250   144.995800    6567.000000  
75%     -37.758200   145.052700   10175.000000  
max     -37.457090   145.526350   21650.000000

Now we would like to predict the Price. The Variable y is used to store the DataFrame Column. 

In [30]:
# prediction_target
y = house_data.Price

Now in order to store features in a Variable. We first must be able to view what are the available options.

In [31]:
house_data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

House Prices are effected by a lot of factors but generally for simplicitity we may select the features to be Rooms, Bathrooms, Landsize, Lattitude and Longtitude.

In [32]:
# features
house_features = ['Rooms','Bathroom','Landsize','Lattitude','Longtitude']

Storing the DataFrame with required features inside Variable X.

In [33]:
X = house_data[house_features]

In [34]:
X.head()

Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954

In [35]:
X.describe()

Rooms     Bathroom      Landsize    Lattitude   Longtitude
count  6196.000000  6196.000000   6196.000000  6196.000000  6196.000000
mean      2.931407     1.576340    471.006940   -37.807904   144.990201
std       0.971079     0.711362    897.449881     0.075850     0.099165
min       1.000000     1.000000      0.000000   -38.164920   144.542370
25%       2.000000     1.000000    152.000000   -37.855438   144.926198
50%       3.000000     1.000000    373.000000   -37.802250   144.995800
75%       4.000000     2.000000    628.000000   -37.758200   145.052700
max       8.000000     8.000000  37000.000000   -37.457090   145.526350

Now we can import DecisionTreeRegressor and hold it to a random state and use it to train the Model using X and y.

In [36]:
from sklearn.tree import DecisionTreeRegressor

#define model
house_model = DecisionTreeRegressor(random_state=1)

#fit model
house_model.fit(X,y)

DecisionTreeRegressor(random_state=1)

In [37]:
print('Making Prediction for the following 5 Houses are')
print(X.head())
print('The Predictions are', house_model.predict(X.head()))

Making Prediction for the following 5 Houses are
   Rooms  Bathroom  Landsize  Lattitude  Longtitude
1      2       1.0     156.0   -37.8079    144.9934
2      3       2.0     134.0   -37.8093    144.9944
4      4       1.0     120.0   -37.8072    144.9941
6      3       2.0     245.0   -37.8024    144.9993
7      2       1.0     256.0   -37.8060    144.9954
The Predictions are [1035000. 1465000. 1600000. 1876000. 1636000.]


Did you notice something? We train the model and then asked the model to predict the prices using features from the same data set we trained the Model on. How will we check how much efficient our Model is! Is it making any mistake? Well there are many methods to check the efficiency of our Model and one of them is mean_absolute_error

In [17]:
predicted_home_prices = house_model.predict(X)

In [18]:
from sklearn.metrics import mean_absolute_error

In [38]:
mae = mean_absolute_error(y, predicted_home_prices)
mae

1115.7467183128902

On Average, our predictions are off by about $1115.77

Another way of handling this issue is to split the data into test data and train data. Use the train data part to train the model and check the efficiency of the model by test data!

In [52]:
from sklearn.model_selection import train_test_split

In [53]:
train_X, val_X, train_y, val_y = train_test_split(X,y, random_state=0)

In [54]:
house_model2 = DecisionTreeRegressor(random_state=1)

In [55]:
house_model2.fit(train_X, train_y)

DecisionTreeRegressor(random_state=1)

In [56]:
predicted_X = house_model2.predict(val_X)

In [57]:
mae2 = mean_absolute_error(val_y, predicted_X)
mae2

273518.01872175594

In [58]:
mae2 - mae

272402.27200344304

Wow, when we train and test Model without splitting the Data the mean absolute error was only \\$1000 and when we trained and tested the model using different sample from the same data, the mean absolute error reached up to \\$272,000